File: app/utils/dawproject/timeline/mediaFile.ts
import { DoubleAdapter } from "../doubleAdapter";
import { FileReference } from "../fileReference";
import type { IFileReference, IMediaFile } from "../types"; // Added IFileReference
import { Utility } from "../utility";
import { Timeline } from "./timeline";
import { TimeUnit } from "./timeUnit";

export abstract class MediaFile extends Timeline implements IMediaFile {
  duration: number;
  file: IFileReference;

  constructor(
    // Make file optional for deserialization, fromXmlObject will set it
    file?: IFileReference,
    duration: number = 0.0,
    name?: string,
    timeUnit?: TimeUnit
  ) {
    super(undefined, timeUnit, name); // MediaFile doesn't have a track attribute directly
    // Provide a default placeholder for file
    this.file = file || new FileReference("");
    this.duration = duration;
  }

  toXmlObject(): any {
    const obj: any = super.toXmlObject(); // get attributes from Timeline

    // add required duration attribute
    Utility.addAttribute(obj, "duration", this, {
      required: true,
      adapter: DoubleAdapter.toXml,
    });

    // handle the 'File' child directly
    if (this.file && "toXmlObject" in this.file) {
      obj.File = (this.file as FileReference).toXmlObject();
    } else if (this.file) {
      // handle case where file is IFileReference but not FileReference
      console.warn(
        "File property is IFileReference but not FileReference instance. Cannot call toXmlObject."
      );
    }

    // since MediaFile is abstract, it doesn't return a root element itself.
    // subclasses will wrap these attributes and children in their specific root tag.
    return obj;
  }

  fromXmlObject(xmlObject: any): this {
    super.fromXmlObject(xmlObject); // populate inherited attributes from Timeline

    // validate and populate required time attribute
    Utility.populateAttribute<number>(xmlObject, "duration", this, {
      required: true,
      adapter: DoubleAdapter.fromXml,
    });

    if (xmlObject.File) {
      this.file = new FileReference().fromXmlObject(xmlObject.File);
    } else {
      // Handle missing required file
      console.warn(
        "Missing required 'File' element in XML for MediaFile. Assigning default."
      );
      this.file = new FileReference(""); // Assign a default or throw error
    }
    return this;
  }
}

File: app/utils/dawproject_python/mediaFile.py
from lxml import etree as ET
from classes.fileReference import FileReference
from classes.timeline import Timeline


class MediaFile(Timeline):
    def __init__(self, file=None, duration=0.0, name=None):
        super().__init__(name)  # Call the base class (Timeline) constructor
        self.file = (
            file if file else FileReference()
        )  # Ensure file is a FileReference instance
        self.duration = duration

    def to_xml(self):
        # The tag name will be determined by the subclass, e.g., "Audio", "Video"
        media_elem = ET.Element(
            "MediaFile"
        )  # Placeholder, should be overridden by subclass

        # Add the File element
        file_elem = self.file.to_xml()
        file_elem.tag = "File"  # Ensure the tag is "File"
        media_elem.append(file_elem)

        # Set the duration attribute
        media_elem.set("duration", str(self.duration))

        return media_elem

    @classmethod
    def from_xml(cls, element):
        # Parse the File element
        file_elem = element.find("File")
        file = (
            FileReference.from_xml(file_elem)
            if file_elem is not None
            else FileReference()
        )

        # Parse the duration attribute
        duration = float(element.get("duration", 0.0))

        # Optional name attribute from Timeline
        name = element.get("name")

        return cls(file=file, duration=duration, name=name)
